### ISSUE TITLE
[Question] Dynamically adding filter.

### ISSUE THREAD

**Issue by vxhviet** [wysaid/android-gpuimage-plus#178]
In GPUImage, there is [`GPUImageFilterGroup`](https://github.com/CyberAgent/android-gpuimage/blob/master/library/src/jp/co/cyberagent/android/gpuimage/GPUImageFilterGroup.java) that let us stack filters as we like. I would like to ask is there anything similar in GPU Image Plus, as from your examples, I see you always predefine it at first, something like this:

`"@adjust brightness 0 @adjust contrast 1 @adjust saturation 1 @adjust sharpen 0 @adjust lut filter_lut_bw_12.png"`

I would like to dynamically add filter based on user input, so it can either be `"@adjust brightness 0 @adjust contrast 1"` or `"@adjust contrast 1 @adjust saturation 1 @adjust sharpen"`.

I ask this because in GPUImage it very inefficient to process multiple filters at the same time, so dynamically adding them one by one will give a better experience. GPUImage Plus after some testing is very efficient, so I don't know if this is necessary. But I still would like to dynamically adding them just in case the number of filter increase.

**Comment by wysaid** [comment]
You may need [this](https://github.com/wysaid/android-gpuimage-plus/blob/master/library/src/main/jni/include/cgeImageHandler.h)
You can use a CGEImageHandler to do that at JNI layer.

**Comment by vxhviet** [comment]
Thank you very much for your response. I'll see what I can do. I'm not familiar with C++ though.

Is there any chance we get that supported in the next release?

**Comment by wysaid** [comment]
There's a java wrapper for that: https://github.com/wysaid/android-gpuimage-plus/blob/master/library/src/main/java/org/wysaid/nativePort/CGEImageHandler.java

**Comment by vxhviet** [comment]
Thanks, I've solved this by dynamically manipulate the config string. So basically just create a new config String depending on the input. GPUImagePlus is indeed very efficient so I won't notice any dramatically performance hit so far. That said, I'm only manipulate 5 filters at the moment (LUT, contrast, brightness, exposure, saturation) so I won't really know if this approach is really the best for say 50+ filters.

Based on my observation, dynamically adding `@adjust lut` seem to have a problem with not immediately update the content for the first time. I've try a variety of these:

```
        mGLSurfaceView.requestRender();
        mGLSurfaceView.invalidate();
        mGLSurfaceView.postInvalidate();
        mGLSurfaceView.requestLayout();
        mGLSurfaceView.getImageHandler().processFilters();
        mGLSurfaceView.getImageHandler().drawResult();
```

But none seem to work. This is just a minor issue so I've decided to always prefix the config string with a neutral `@adjust lut`. Work great for my purpose.
